### PR TITLE
Add isin()

### PR DESCRIPTION
**Pull request by garawalid**
`isin()` method for series and dataframe.

### CODE REVIEW

**Comment by rxin** [comment]
Can you break this into two separate PRs?

One for isin, and the other for values / get_values. I'm not sure if we should put values and get_values there, because it might blow up memory. One way we can do this is for each function that requires collecting all the data to a single node, we require users to run toPandas explicitly, so they would know what they are doing and not shoot themselves in the foot.

**Comment by garawalid** [comment]
I divided this PR into two.

**Comment by thunterdb** [comment]
@garawalid thanks for the PR. Would you mind resolving the conflicts?

**Comment by rxin** [comment]
@makimat - do you have a bot that goes around and add that visualization randomly?

File: databricks/koalas/missing/series.py [hunk 1]
```diff
@@ -87,15 +87,13 @@ class _MissingPandasLikeSeries(object):
     get_dtype_counts = unsupported_function('get_dtype_counts')
     get_ftype_counts = unsupported_function('get_ftype_counts')
     get_value = unsupported_function('get_value')
-    get_values = unsupported_function('get_values')
```
Comment (rxin) [inline databricks/koalas/missing/series.py]: need to undo this change

Comment (garawalid) [reply databricks/koalas/missing/series.py]: Yes, you are right! 
I forget these changes in both `series.py` and `frame.py`.
@rxin  Thanks!

**Comment by garawalid** [comment]
@thunterdb I resolved the conflict.

**Comment by makimat** [comment]
> @makimat - do you have a bot that goes around and add that visualization randomly?

It doesn't do it randomly - your repo looked interesting and I wanted to visualize this specific pull request so I asked our bot to do it. You can do it for any pr out there at https://softagram.com/pull-request-bot. Let me know what do you think about the visualization!

**Comment by rxin** [comment]
Thanks @makimat. The viz looks interesting. We might try this!

File: databricks/koalas/series.py [hunk 2]
```diff
@@ -321,6 +321,24 @@ def value_counts(self, normalize=False, sort=True, ascending=False, bins=None, d
         kdf._metadata = Metadata(column_fields=[self.name], index_info=[(index_name, None)])
         return _col(kdf)
 
+    @derived_from(pd.Series)
```
Comment (thunterdb) [inline databricks/koalas/series.py]: Since you put documentation already, you do not need this annotation (which is just here to inject the documentation from pandas).

File: databricks/koalas/tests/test_dataframe.py [hunk 3]
```diff
@@ -360,10 +360,43 @@ def test_missing(self):
...
+
+        msg = "Dataframe and Series are not supported"
+        with self.assertRaisesRegex(NotImplementedError, msg):
+            ddf.isin(df)
+
+    def test_isin_series(self):
+
+        s = pd.Series(['lama', 'cow', 'lama', 'beetle', 'lama', 'hippo'], name='animal')
+
+        ds = koalas.from_pandas(pd.DataFrame(s)).animal
```
Comment (thunterdb) [inline databricks/koalas/tests/test_dataframe.py]: you should be able to do `ks.Series(s)` now.

File: databricks/koalas/frame.py [hunk 4]
```diff
@@ -400,6 +403,40 @@ def groupby(self, by):
         gp = self._sdf.groupby(by)
         return PandasLikeGroupBy(self, gp, None)
 
+    @derived_from(pd.DataFrame)
```
Comment (thunterdb) [inline databricks/koalas/frame.py]: see comment below

File: databricks/koalas/frame.py [hunk 5]
```diff
@@ -22,8 +22,11 @@
 
 import numpy as np
 import pandas as pd
+
 from pyspark import sql as spark
-from pyspark.sql import functions as F
+from collections.abc import Iterable
+from pyspark.sql import Column, DataFrame, functions as F
```
Comment (ueshin) [inline databricks/koalas/frame.py]: We should use `spark.Column` or `spark.DataFrame` instead of importing from `pyspark` since the names conflict with Koalas' ones.
Btw, they are not used?

File: databricks/koalas/frame.py [hunk 6]
```diff
@@ -400,6 +403,40 @@ def groupby(self, by):
...
+                    "'DataFrame' object has no attribute %s"
+                    % (set(values.keys()).difference(self.columns)))
+        elif not isinstance(values, (pd.DataFrame, pd.Series)) and isinstance(values, Iterable):
+            _select_columns = self._metadata.index_fields + [self[col]._scol.isin(values).alias(col)
+                                                             for col in self.columns]
+            return DataFrame(self._sdf.select(_select_columns), self._metadata.copy())
+        elif isinstance(values, (pd.DataFrame, pd.Series)):
+            raise NotImplementedError("Dataframe and Series are not supported")
+        else:
+            TypeError('Values should be iterable, Series, DataFrame or dict.')
```
Comment (ueshin) [inline databricks/koalas/frame.py]: We need `raise` ?

File: databricks/koalas/tests/test_dataframe.py (hunk 3 above)
Comment (ueshin) [reply databricks/koalas/tests/test_dataframe.py]: Yes, or `koalas.from_pandas(s)` should work as well.

File: databricks/koalas/frame.py [hunk 7]
```diff
@@ -22,8 +22,11 @@
 
 import numpy as np
 import pandas as pd
+
 from pyspark import sql as spark
-from pyspark.sql import functions as F
+from collections.abc import Iterable
+from pyspark.sql import Column, DataFrame, functions as F
+
```
Comment (ueshin) [inline databricks/koalas/frame.py]: Please remove the line in the same category import block.

File: databricks/koalas/frame.py [hunk 8]
```diff
@@ -22,8 +22,11 @@
 
 import numpy as np
 import pandas as pd
+
 from pyspark import sql as spark
-from pyspark.sql import functions as F
+from collections.abc import Iterable
```
Comment (ueshin) [inline databricks/koalas/frame.py]: `collections.abc` should be placed before `decorator`.
Basically we put standard libraries on the top import block.

**Comment by ueshin** [comment]
Also could you add docs at the corresponding places in `docs/source/reference/frame.rst` and `docs/sources/references/series.rst`?
Thanks!

File: databricks/koalas/frame.py [hunk 9]
```diff
@@ -400,6 +403,40 @@ def groupby(self, by):
...
+    def isin(self, values):
+        """
+        Check whether values are contained in Series.
+
+        :param values: iterable or dict
+                       The sequence of values to test. If values is a dict,
+                       the keys must be the column names, which must match.
+                       Series and DataFrame are not supported.
+        :return: :class:`DataFrame`
+        """
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: and doc test + inlining with Pandas doc (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html#pandas.DataFrame.isin)

**Comment by HyukjinKwon** [comment]
@garawalid, mind updating please? Looks getting close.

File: databricks/koalas/frame.py (hunk 8 above)
Comment (garawalid) [reply databricks/koalas/frame.py]: Thanks !

File: databricks/koalas/frame.py (hunk 6 above)
Comment (garawalid) [reply databricks/koalas/frame.py]: Agree.

File: databricks/koalas/series.py (hunk 2 above)
Comment (garawalid) [reply databricks/koalas/series.py]: You are right! Thanks!

**Comment by garawalid** [comment]
@HyukjinKwon 
I updated the PR. 

@thunterdb 
I created a `list_sanitizer()` to assure data sanitization in a given list.

File: databricks/koalas/frame.py [hunk 10]
```diff
@@ -947,6 +948,63 @@ def groupby(self, by):
         gp = self._sdf.groupby(by)
         return PandasLikeGroupBy(self, gp, None)
 
+    def isin(self, values):
+        """
+        Whether each element in the DataFrame is contained in values.
+
+        :param values: iterable or dict
```
Comment (rxin) [inline databricks/koalas/frame.py]: can you just use the numpy docs format, similar to Pandas?

https://github.com/pandas-dev/pandas/blob/v0.24.2/pandas/core/frame.py#L7857-L7944

File: databricks/koalas/frame.py [hunk 11]
```diff
@@ -947,6 +948,63 @@ def groupby(self, by):
...
+                       Series and DataFrame are not supported.
+        :return: :class:`DataFrame`
+                        DataFrame of booleans showing whether each element
+                        in the DataFrame is contained in values.
+
+        Examples
+        --------
+        When ``values`` is a list:
+
+        >>> df = ks.DataFrame({'A': [1, 2, 3], 'B': ['a', 'b', 'f']})
```
Comment (rxin) [inline databricks/koalas/frame.py]: why not use the same examples in the pandas doc?

File: databricks/koalas/tests/test_dataframe.py [hunk 12]
```diff
@@ -426,3 +426,54 @@ def test_to_pandas(self):
...
+        # Test list sanitizer
+        msg = "List contains unsupported type <class 'pandas.core.frame.DataFrame'>"
+        with self.assertRaisesRegex(TypeError, msg):
+            kdf.isin([df, df])
+
+        msg = "List contains unsupported type <class 'pandas.core.frame.DataFrame'>"
+        with self.assertRaisesRegex(TypeError, msg):
+            kdf.isin({"a": [5, df]})
+
+    def test_isin_series(self):
```
Comment (rxin) [inline databricks/koalas/tests/test_dataframe.py]: I'd create test_series.py file and move this there.

File: databricks/koalas/frame.py (hunk 11 above)
Comment (garawalid) [reply databricks/koalas/frame.py]: I took it from the source code of my Pandas 0.23.4 
I'll change it by the latest one.

File: databricks/koalas/tests/test_dataframe.py (hunk 12 above)
Comment (garawalid) [reply databricks/koalas/tests/test_dataframe.py]: So, we move also these methods: `test_Series()`, `test_rename_series()`, `test_rename_series_method()` to `test_series.py` ?

File: databricks/koalas/frame.py (hunk 10 above)
Comment (garawalid) [reply databricks/koalas/frame.py]: Okay! 
btw, many methods use restructuredtext style  like `reset_index()`.

File: databricks/koalas/tests/test_dataframe.py (hunk 12 above)
Comment (rxin) [reply databricks/koalas/tests/test_dataframe.py]: You can create a new file and just put this test case there, and create a new pull request in the future to move the rest if you want. In a way that's better because the commits are more atomic and self-contained.

Comment (garawalid) [reply databricks/koalas/tests/test_dataframe.py]: Sounds good!

**Comment by garawalid** [comment]
cc @rxin

**Comment by HyukjinKwon** [comment]
@garawalid, can you rebase this please? Let me take a final pass and merge

**Comment by garawalid** [comment]
@HyukjinKwon 
I updated the PR.

**Comment by HyukjinKwon** [comment]
Merged - it had many review iterations. Let me make a followup to clean up.
